// @aws-quickstart/eks-blueprints/dist/resource-providers/ipv6-vpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ipv6VpcProvider = void 0;
const aws_cdk_lib_1 = require("aws-cdk-lib");
const ec2 = require("aws-cdk-lib/aws-ec2");
const aws_ec2_1 = require("aws-cdk-lib/aws-ec2");
const vpc_1 = require("./vpc");
/**
 * IPV6 VPC resource provider
 */
class Ipv6VpcProvider {
    constructor(vpcId) {
        this.vpcId = vpcId;
    }
    provide(context) {
        const id = context.scope.node.id;
        let vpc = (0, vpc_1.getVPCFromId)(context, id, this.vpcId);
        if (vpc == null) {
            // It will automatically divide the provided VPC CIDR range, and create public and private subnets per Availability Zone.
            // Network routing for the public subnets will be configured to allow outbound access directly via an Internet Gateway.
            // Network routing for the private subnets will be configured to allow outbound access via a one NAT Gateway to reduce the cost.
            // IPv6 does not require NAT for pod to pod communication. By default, we are creating one NAT for cluster communications outside endpoints if any.
            return this.getIPv6VPC(context, id);
        }
        return vpc;
    }
    /*
    ** AWS recommend to have dual stack vpc for ipv6 EKS clusters. This functions creates VPC required for IPV6 cluster.
    ** For more info refer: https://docs.aws.amazon.com/vpc/latest/userguide/vpc-migrate-ipv6-add.html
     */
    getIPv6VPC(context, id) {
        // Create VPC with dual stack mode
        // Setting natGateways lower than the number of Availability Zones in VPC in order to save on NAT cost.
        const vpc = new ec2.Vpc(context.scope, id + "-vpc", { natGateways: 1,
            ipProtocol: aws_ec2_1.IpProtocol.DUAL_STACK, restrictDefaultSecurityGroup: false });
        // Create and associate IPV6 CIDR blocks
        const ipv6Cidr = new ec2.CfnVPCCidrBlock(context.scope, id + "-CIDR6", {
            vpcId: vpc.vpcId,
            amazonProvidedIpv6CidrBlock: true,
        });
        let subnetCount = 0;
        let subnets = [...vpc.publicSubnets, ...vpc.privateSubnets];
        // associate an IPv6 CIDR block with a subnet
        for (let subnet of subnets) {
            // Wait for the ipv6 cidr to complete
            subnet.node.addDependency(ipv6Cidr);
            this.associateSubnetsWithIpv6CIDR(subnetCount, subnet, vpc);
            subnetCount++;
        }
        return vpc;
    }
    /*
    ** For IPV6 vpc we need to attach subnets with available ipv6Cidr blocks in the vpc.
    ** Refer steps in here: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-cidr.html
     */
    associateSubnetsWithIpv6CIDR(count, subnet, vpc) {
        const cfnSubnet = subnet.node.defaultChild;
        // The VPC is associated with /56 for amazonProvidedIpv6CidrBlock. So value of 64 subnet mask. so 256 cidr blocks are available.
        // Having 64 as subnet mask will give 2^64 IP's for each subnet. Which high enough for any kind of workload.
        const ipv6CIDRSubnetMask = "64";
        cfnSubnet.ipv6CidrBlock = aws_cdk_lib_1.Fn.select(count, aws_cdk_lib_1.Fn.cidr(aws_cdk_lib_1.Fn.select(0, vpc.vpcIpv6CidrBlocks), 256, ipv6CIDRSubnetMask));
        cfnSubnet.assignIpv6AddressOnCreation = true;
    }
}
exports.Ipv6VpcProvider = Ipv6VpcProvider;
