// @aws-quickstart/eks-blueprints/dist/resource-providers/ipv6-iam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateIPv6NodeRoleProvider = void 0;
const iam = require("aws-cdk-lib/aws-iam");
const ipv6_utils_1 = require("../utils/ipv6-utils");
/**
 * Resource provider that creates a new role with ipv6 permissions.
 * Especially, Node management roles (requires ipv6 permissions).
 */
class CreateIPv6NodeRoleProvider {
    /**
     * Constructor to create role provider.
     * @param roleId role id
     * @param assumedBy @example  new iam.ServicePrincipal('ec2.amazonaws.com')
     * @param policies
     */
    constructor(roleId, policies) {
        this.roleId = roleId;
        this.policies = policies;
    }
    provide(context) {
        const assumedBy = new iam.ServicePrincipal("ec2.amazonaws.com");
        const defaultNodePolicies = [
            iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonEKSWorkerNodePolicy"),
            iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonEC2ContainerRegistryReadOnly"),
            iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonSSMManagedInstanceCore")
        ];
        const nodePolicies = this.policies ? defaultNodePolicies.concat(this.policies) : defaultNodePolicies;
        const role = new iam.Role(context.scope, this.roleId, {
            assumedBy: assumedBy,
            managedPolicies: nodePolicies
        });
        const nodeIpv6Policy = new iam.Policy(context.scope, 'node-Ipv6-Policy', {
            document: (0, ipv6_utils_1.getEKSNodeIpv6PolicyDocument)()
        });
        role.attachInlinePolicy(nodeIpv6Policy);
        return role;
    }
}
exports.CreateIPv6NodeRoleProvider = CreateIPv6NodeRoleProvider;
