// @aws-quickstart/eks-blueprints/dist/addons/eks-pod-identity-agent/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EksPodIdentityAgentAddOn = void 0;
const core_addon_1 = require("../core-addon");
const aws_eks_1 = require("aws-cdk-lib/aws-eks");
const utils = require("../../utils");
const versionMap = new Map([
    [aws_eks_1.KubernetesVersion.V1_31, "v1.3.4-eksbuild.1"],
    [aws_eks_1.KubernetesVersion.V1_30, "v1.3.2-eksbuild.2"],
    [aws_eks_1.KubernetesVersion.V1_29, "v1.3.2-eksbuild.2"],
    [aws_eks_1.KubernetesVersion.V1_28, "v1.3.2-eksbuild.2"],
    [aws_eks_1.KubernetesVersion.V1_27, "v1.3.2-eksbuild.2"],
    [aws_eks_1.KubernetesVersion.V1_26, "v1.3.2-eksbuild.2"],
]);
/**
 * Default values for the add-on
 */
const defaultProps = {
    addOnName: 'eks-pod-identity-agent',
    version: 'auto',
    versionMap: versionMap,
    saName: "eks-pod-identity-agent-sa",
};
/**
 * Implementation of Amazon EKS Pod Identity Agent add-on.
 */
class EksPodIdentityAgentAddOn extends core_addon_1.CoreAddOn {
    deploy(clusterInfo) {
        return super.deploy(clusterInfo);
    }
    constructor(version) {
        super({
            addOnName: defaultProps.addOnName,
            version: version !== null && version !== void 0 ? version : defaultProps.version,
            saName: defaultProps.saName,
            versionMap: defaultProps.versionMap
        });
    }
}
exports.EksPodIdentityAgentAddOn = EksPodIdentityAgentAddOn;
__decorate([
    utils.conflictsWithAutoMode('v1.3.4-eksbuild.1')
], EksPodIdentityAgentAddOn.prototype, "deploy", null);
