// @aws-quickstart/eks-blueprints/dist/addons/adot/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdotCollectorAddOn = void 0;
const aws_eks_1 = require("aws-cdk-lib/aws-eks");
const utils_1 = require("../../utils");
const cert_manager_1 = require("../cert-manager");
const core_addon_1 = require("../core-addon");
const iam_policy_1 = require("./iam-policy");
const aws_eks_2 = require("aws-cdk-lib/aws-eks");
const versionMap = new Map([
    [aws_eks_2.KubernetesVersion.V1_31, "v0.109.0-eksbuild.2"],
    [aws_eks_2.KubernetesVersion.V1_30, "v0.109.0-eksbuild.2"],
    [aws_eks_2.KubernetesVersion.V1_29, "v0.109.0-eksbuild.2"],
    [aws_eks_2.KubernetesVersion.V1_28, "v0.109.0-eksbuild.2"],
    [aws_eks_2.KubernetesVersion.V1_27, "v0.109.0-eksbuild.2"],
    [aws_eks_2.KubernetesVersion.V1_26, "v0.109.0-eksbuild.2"],
]);
const defaultProps = {
    addOnName: 'adot',
    version: 'auto',
    versionMap: versionMap,
    saName: 'adot-collector',
    policyDocumentProvider: iam_policy_1.getAdotCollectorPolicyDocument,
    namespace: 'default',
    configurationValues: {}
};
/**
 * Implementation of Adot Collector EKS add-on.
 */
let AdotCollectorAddOn = class AdotCollectorAddOn extends core_addon_1.CoreAddOn {
    constructor(props) {
        var _a, _b;
        super({
            ...defaultProps,
            ...props,
            namespace: (_a = props === null || props === void 0 ? void 0 : props.namespace) !== null && _a !== void 0 ? _a : defaultProps.namespace,
            version: (_b = props === null || props === void 0 ? void 0 : props.version) !== null && _b !== void 0 ? _b : defaultProps.version
        });
    }
    deploy(clusterInfo) {
        const addOnPromise = super.deploy(clusterInfo);
        return addOnPromise;
    }
    /**
     * Overriding base class method to create namespace and register permissions.
     * @param clusterInfo
     * @param name
     * @returns
     */
    createNamespace(clusterInfo, namespaceName) {
        // Create namespace if not default
        const cluster = clusterInfo.cluster;
        const ns = (0, utils_1.createNamespace)(namespaceName, cluster, true, true);
        // Applying ADOT Permission manifest
        const otelPermissionsDoc = (0, utils_1.readYamlDocument)(__dirname + '/otel-permissions.yaml');
        const otelPermissionsManifest = otelPermissionsDoc.split("---").map(e => (0, utils_1.loadYaml)(e));
        const otelPermissionsStatement = new aws_eks_1.KubernetesManifest(cluster.stack, "adot-addon-otelPermissions", {
            cluster,
            manifest: otelPermissionsManifest,
            overwrite: true,
        });
        otelPermissionsStatement.node.addDependency(ns);
        return otelPermissionsStatement;
    }
};
exports.AdotCollectorAddOn = AdotCollectorAddOn;
__decorate([
    (0, utils_1.dependable)(cert_manager_1.CertManagerAddOn.name)
], AdotCollectorAddOn.prototype, "deploy", null);
exports.AdotCollectorAddOn = AdotCollectorAddOn = __decorate([
    utils_1.supportsALL
], AdotCollectorAddOn);
