// @aws-quickstart/eks-blueprints/dist/builders/automode-builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutomodeBuilder = void 0;
const cluster_providers_1 = require("../cluster-providers");
const eks = require("aws-cdk-lib/aws-eks");
const ts_deepmerge_1 = require("ts-deepmerge");
const stacks_1 = require("../stacks");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const addons = require("../addons");
const utils = require("../utils");
const defaultOptions = {
    version: eks.KubernetesVersion.V1_31,
    nodePools: ['system', 'general-purpose']
};
class AutomodeBuilder extends stacks_1.BlueprintBuilder {
    static builder(options) {
        const builder = new AutomodeBuilder();
        const mergedOptions = (0, ts_deepmerge_1.merge)(defaultOptions, options);
        builder
            .clusterProvider(new cluster_providers_1.AutomodeClusterProvider(mergedOptions))
            .addOns(new addons.NestedStackAddOn({
            id: "usage-tracking-addon",
            builder: UsageTrackingAddOn.builder(),
        }));
        return builder;
    }
}
exports.AutomodeBuilder = AutomodeBuilder;
/**
 * Nested stack that is used as tracker for GPU Accelerator
 */
class UsageTrackingAddOn extends aws_cdk_lib_1.NestedStack {
    static builder() {
        return {
            build(scope, id, props) {
                return new UsageTrackingAddOn(scope, id, props);
            }
        };
    }
    constructor(scope, id, props) {
        super(scope, id, utils.withUsageTracking(UsageTrackingAddOn.USAGE_ID, props));
    }
}
UsageTrackingAddOn.USAGE_ID = "SOME_VALUE";
