// @aws-quickstart/eks-blueprints/dist/utils/namespace-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createNamespace = createNamespace;
const aws_eks_1 = require("aws-cdk-lib/aws-eks");
/**
  * Creates namespace
  * (a prerequisite for serviceaccount and helm chart execution for many add-ons).
  * @param name
  * @param cluster
  * @param overwrite
  * @param prune
  * @returns KubernetesManifest
  */
function createNamespace(name, cluster, overwrite, prune, annotations, labels) {
    if (name === "kube-system") {
        return cluster.clusterSecurityGroup; // a construct that is populated for the cluster l
    }
    return new aws_eks_1.KubernetesManifest(cluster.stack, `${name}-namespace-struct`, {
        cluster: cluster,
        manifest: [{
                apiVersion: 'v1',
                kind: 'Namespace',
                metadata: {
                    name: name,
                    annotations,
                    labels
                }
            }],
        overwrite: overwrite !== null && overwrite !== void 0 ? overwrite : true,
        prune: prune !== null && prune !== void 0 ? prune : true
    });
}
