// @aws-quickstart/eks-blueprints/dist/utils/yaml-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyYamlFromDir = applyYamlFromDir;
exports.readYamlDocument = readYamlDocument;
exports.loadMultiResourceYaml = loadMultiResourceYaml;
exports.loadYaml = loadYaml;
exports.loadExternalYaml = loadExternalYaml;
exports.serializeYaml = serializeYaml;
exports.convertKeyPair = convertKeyPair;
const fs = require("fs");
const yaml = require("js-yaml");
/**
 * Applies all manifests from a directory. Note: The manifests are not checked,
 * so user must ensure the manifests have the correct namespaces.
 * @param dir
 * @param cluster
 * @param namespaceManifest
 */
function applyYamlFromDir(dir, cluster, namespaceManifest) {
    fs.readdirSync(dir, { encoding: 'utf8' }).forEach((file, index) => {
        if (file.split('.').pop() == 'yaml') {
            const data = fs.readFileSync(dir + file, 'utf8');
            if (data != undefined) {
                yaml.loadAll(data, function (item) {
                    const resources = cluster.addManifest(file.substring(0, file.length - 5) + index, item);
                    resources.node.addDependency(namespaceManifest);
                });
            }
        }
    });
}
/**
 * Reads the YAML document from a local path.
 * @param path YAML document path
 * @returns YAML document string
 */
function readYamlDocument(path) {
    try {
        const doc = fs.readFileSync(path, 'utf8');
        return doc;
    }
    catch (e) {
        console.log(e + ' for path: ' + path);
        throw e;
    }
}
/**
 * Reads the YAML document from a local path and parses them as
 * multiple YAML documents separated by `---` as expected in a Kubernetes manifest file
 * @param path YAML document path
 * @returns a list of parsed YAML documents
 */
function loadMultiResourceYaml(path) {
    const doc = readYamlDocument(path);
    return doc.split("---").map((e) => loadYaml(e));
}
/**
 * Parses the sting document into a single YAML document
 * @param document document
 * @returns yaml document
 */
function loadYaml(document) {
    return yaml.load(document);
}
/**
 * Reads the YAML document from a URL and parses
 * multiple YAML documents separated by `---` as expected in a Kubernetes manifest file Note: The file from the URL is
 * not validated, so user must ensure the URL contains a valid manifest.
 * @param url YAML document URL
 * @returns a list of parsed YAML documents
 */
function loadExternalYaml(url) {
    /* eslint-disable */
    const request = require('sync-request'); // moved away from import as it is causing open handles that prevents jest from completion
    const response = request('GET', url);
    return yaml.loadAll(response.getBody().toString());
}
/**
 * Serializes object as a YAML document
 * @param document document
 * @returns yaml document
 */
function serializeYaml(document) {
    return yaml.dump(document);
}
/**
 * Helper function to convert a key-pair values (with an operator)
 * of spec configurations to appropriate json format for addManifest function
 * @param reqs
 * @returns newReqs
 * */
function convertKeyPair(reqs) {
    const newReqs = [];
    for (let req of reqs) {
        const key = req["key"];
        const op = req["operator"];
        const val = req["values"];
        const requirement = { key: key, operator: op, values: val };
        newReqs.push(requirement);
    }
    return newReqs;
}
