// @aws-cdk/aws-eks-v2-alpha/lib/addon.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Addon = void 0;
const jsiiDeprecationWarnings = require("../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_eks_1 = require("aws-cdk-lib/aws-eks");
const core_1 = require("aws-cdk-lib/core");
const metadata_resource_1 = require("aws-cdk-lib/core/lib/metadata-resource");
/**
 * Represents an Amazon EKS Add-On.
 * @resource AWS::EKS::Addon
 */
class Addon extends core_1.Resource {
    /**
     * Creates an `IAddon` instance from the given addon attributes.
     *
     * @param scope - The parent construct.
     * @param id - The construct ID.
     * @param attrs - The attributes of the addon, including the addon name and the cluster name.
     * @returns An `IAddon` instance.
     */
    static fromAddonAttributes(scope, id, attrs) {
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_eks_v2_alpha_AddonAttributes(attrs);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, this.fromAddonAttributes);
            }
            throw error;
        }
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.addonName = attrs.addonName;
                this.addonArn = core_1.Stack.of(scope).formatArn({
                    service: 'eks',
                    resource: 'addon',
                    resourceName: `${attrs.clusterName}/${attrs.addonName}`,
                });
            }
        }
        return new Import(scope, id);
    }
    /**
     * Creates an `IAddon` from an existing addon ARN.
     *
     * @param scope - The parent construct.
     * @param id - The ID of the construct.
     * @param addonArn - The ARN of the addon.
     * @returns An `IAddon` implementation.
     */
    static fromAddonArn(scope, id, addonArn) {
        const parsedArn = core_1.Stack.of(scope).splitArn(addonArn, core_1.ArnFormat.COLON_RESOURCE_NAME);
        const splitResourceName = core_1.Fn.split('/', parsedArn.resourceName);
        class Import extends core_1.Resource {
            constructor() {
                super(...arguments);
                this.addonName = core_1.Fn.select(1, splitResourceName);
                this.addonArn = addonArn;
            }
        }
        return new Import(scope, id);
    }
    /**
     * Creates a new Amazon EKS Add-On.
     * @param scope The parent construct.
     * @param id The construct ID.
     * @param props The properties for the Add-On.
     */
    constructor(scope, id, props) {
        super(scope, id, {
            physicalName: props.addonName,
        });
        try {
            jsiiDeprecationWarnings._aws_cdk_aws_eks_v2_alpha_AddonProps(props);
        }
        catch (error) {
            if (process.env.JSII_DEBUG !== "1" && error.name === "DeprecationError") {
                Error.captureStackTrace(error, Addon);
            }
            throw error;
        }
        // Enhanced CDK Analytics Telemetry
        (0, metadata_resource_1.addConstructMetadata)(this, props);
        this.clusterName = props.cluster.clusterName;
        this.addonName = props.addonName;
        const resource = new aws_eks_1.CfnAddon(this, 'Resource', {
            addonName: props.addonName,
            clusterName: this.clusterName,
            addonVersion: props.addonVersion,
            preserveOnDelete: props.preserveOnDelete,
        });
        this.addonName = this.getResourceNameAttribute(resource.ref);
        this.addonArn = this.getResourceArnAttribute(resource.attrArn, {
            service: 'eks',
            resource: 'addon',
            resourceName: `${this.clusterName}/${this.addonName}/`,
        });
    }
}
exports.Addon = Addon;
_a = JSII_RTTI_SYMBOL_1;
Addon[_a] = { fqn: "@aws-cdk/aws-eks-v2-alpha.Addon", version: "2.187.0-alpha.0" };
