// @aws-quickstart/eks-blueprints/dist/addons/secrets-store/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsStoreAddOn = void 0;
const csi_driver_provider_aws_1 = require("./csi-driver-provider-aws");
const utils_1 = require("../../utils");
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    ascpUrl: 'https://raw.githubusercontent.com/aws/secrets-store-csi-driver-provider-aws/secrets-store-csi-driver-provider-aws-0.3.11/deployment/aws-provider-installer.yaml',
    chart: 'secrets-store-csi-driver',
    name: 'secrets-store-csi-driver',
    namespace: 'kube-system',
    version: '1.5.0',
    release: 'blueprints-addon-secret-store-csi-driver',
    repository: 'https://kubernetes-sigs.github.io/secrets-store-csi-driver/charts',
    rotationPollInterval: undefined,
    syncSecrets: true,
};
let SecretsStoreAddOn = class SecretsStoreAddOn {
    constructor(props) {
        this.options = (0, utils_1.cloneDeep)({ ...defaultProps, ...props });
    }
    deploy(clusterInfo) {
        const csiDriverProviderAws = new csi_driver_provider_aws_1.CsiDriverProviderAws(this.options);
        return Promise.resolve(csiDriverProviderAws.deploy(clusterInfo));
    }
};
exports.SecretsStoreAddOn = SecretsStoreAddOn;
exports.SecretsStoreAddOn = SecretsStoreAddOn = __decorate([
    utils_1.supportsALL
], SecretsStoreAddOn);
