// @aws-quickstart/eks-blueprints/dist/addons/s3-csi-driver/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3CSIDriverAddOn = void 0;
const helm_addon_1 = require("../helm-addon");
const iam = require("aws-cdk-lib/aws-iam");
const utils_1 = require("../../utils");
const iam_policy_1 = require("./iam-policy");
const S3_CSI_DRIVER_SA = 's3-csi-driver-sa';
const S3_CSI_DRIVER = "aws-mountpoint-s3-csi-driver";
const S3_CSI_DRIVER_RELEASE = 's3-csi-driver-release';
const S3_DRIVER_POLICY = 's3-csi-driver-policy';
/**
 * Defaults options for the add-on
 */
const defaultProps = {
    chart: S3_CSI_DRIVER,
    name: S3_CSI_DRIVER,
    namespace: "kube-system",
    release: S3_CSI_DRIVER_RELEASE,
    version: "v1.14.1",
    repository: "https://awslabs.github.io/mountpoint-s3-csi-driver",
    createNamespace: false,
    bucketNames: [],
    kmsArns: []
};
let S3CSIDriverAddOn = class S3CSIDriverAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        // Create service account and policy
        const cluster = clusterInfo.cluster;
        const serviceAccount = cluster.addServiceAccount(S3_CSI_DRIVER_SA, {
            name: S3_CSI_DRIVER_SA,
            namespace: this.options.namespace,
        });
        const s3BucketPolicy = new iam.Policy(cluster, S3_DRIVER_POLICY, {
            statements: (0, iam_policy_1.getS3DriverPolicyStatements)(this.options.bucketNames, (_a = this.options.kmsArns) !== null && _a !== void 0 ? _a : [])
        });
        serviceAccount.role.attachInlinePolicy(s3BucketPolicy);
        // Create namespace
        if (this.options.createNamespace) {
            const ns = (0, utils_1.createNamespace)(this.options.namespace, cluster, true);
            serviceAccount.node.addDependency(ns);
        }
        // setup value for helm chart
        const chartValues = populateValues(this.options);
        const s3CsiDriverChart = this.addHelmChart(clusterInfo, chartValues, true, true);
        s3CsiDriverChart.node.addDependency(serviceAccount);
        return Promise.resolve(s3CsiDriverChart);
    }
};
exports.S3CSIDriverAddOn = S3CSIDriverAddOn;
exports.S3CSIDriverAddOn = S3CSIDriverAddOn = __decorate([
    utils_1.supportsALL
], S3CSIDriverAddOn);
function populateValues(helmOptions) {
    var _a;
    const values = (_a = helmOptions.values) !== null && _a !== void 0 ? _a : {};
    (0, utils_1.setPath)(values, 'node.serviceAccount.create', false);
    (0, utils_1.setPath)(values, 'node.tolerateAllTaints', true);
    return values;
}
