// @aws-quickstart/eks-blueprints/dist/cluster-providers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ISOLATED_CLUSTER = exports.PRIVATE_CLUSTER = exports.DESIRED_SIZE_KEY = exports.MAX_SIZE_KEY = exports.MIN_SIZE_KEY = exports.INSTANCE_TYPE_KEY = exports.DEFAULT_NG_MAXSIZE = exports.DEFAULT_NG_MINSIZE = exports.DEFAULT_INSTANCE_TYPE = void 0;
const ec2 = require("aws-cdk-lib/aws-ec2");
/**
 * Default instance type for managed node group provisioning
 */
exports.DEFAULT_INSTANCE_TYPE = ec2.InstanceType.of(ec2.InstanceClass.M5, ec2.InstanceSize.LARGE);
/**
 * Default min size of MNG
 */
exports.DEFAULT_NG_MINSIZE = 1;
/**
 * Default max size for MNG
 */
exports.DEFAULT_NG_MAXSIZE = 3;
/**
 * Keys for context lookups.
 */
exports.INSTANCE_TYPE_KEY = "eks.default.instance-type";
exports.MIN_SIZE_KEY = "eks.default.min-size";
exports.MAX_SIZE_KEY = "eks.default.max-size";
exports.DESIRED_SIZE_KEY = "eks.default.desired-size";
exports.PRIVATE_CLUSTER = "eks.default.private-cluster";
exports.ISOLATED_CLUSTER = "eks.default.isolated-cluster";
