// @aws-quickstart/eks-blueprints/dist/addons/upbound-universal-crossplane/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var UpboundCrossplaneAddOn_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UpboundCrossplaneAddOn = void 0;
const ts_deepmerge_1 = require("ts-deepmerge");
const utils_1 = require("../../utils");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const cdk = require("aws-cdk-lib");
const helm_addon_1 = require("../helm-addon");
const defaultProps = {
    name: 'uxp',
    release: 'blueprints-addon-uxp',
    namespace: 'upbound-system',
    chart: 'universal-crossplane',
    version: '1.14.5-up.1',
    repository: 'https://charts.upbound.io/stable',
    values: {}
};
let UpboundCrossplaneAddOn = UpboundCrossplaneAddOn_1 = class UpboundCrossplaneAddOn extends helm_addon_1.HelmAddOn {
    constructor(props) {
        super({ ...defaultProps, ...props });
        this.options = this.props;
    }
    deploy(clusterInfo) {
        var _a;
        const cluster = clusterInfo.cluster;
        // Create the `upbound-system` namespace.
        const ns = (0, utils_1.createNamespace)(this.options.namespace, cluster, true);
        // Create the CrossPlane AWS Provider IRSA.
        const serviceAccountName = "provider-aws";
        const sa = cluster.addServiceAccount(serviceAccountName, {
            name: serviceAccountName,
            namespace: this.options.namespace,
        });
        sa.node.addDependency(ns);
        sa.role.attachInlinePolicy(new aws_iam_1.Policy(cluster.stack, 'eks-connect-policy', {
            document: aws_iam_1.PolicyDocument.fromJson({
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Effect": "Allow",
                        "Action": ["sts:AssumeRole"],
                        "Resource": `${this.options.clusterAccessRole.roleArn}`
                    },
                    {
                        "Effect": "Allow",
                        "Action": ["eks:*"],
                        "Resource": `*`
                    }
                ]
            })
        }));
        clusterInfo.addAddOnContext(UpboundCrossplaneAddOn_1.name, {
            arn: sa.role.roleArn
        });
        new cdk.CfnOutput(cluster.stack, 'providerawssaiamrole', {
            value: sa.role.roleArn,
            description: 'provider AWS IAM role',
            exportName: 'providerawssaiamrole'
        });
        let values = (_a = this.options.values) !== null && _a !== void 0 ? _a : {};
        values = (0, ts_deepmerge_1.merge)(values, values);
        const chart = this.addHelmChart(clusterInfo, values, false, true);
        chart.node.addDependency(sa);
        return Promise.resolve(chart);
    }
};
exports.UpboundCrossplaneAddOn = UpboundCrossplaneAddOn;
exports.UpboundCrossplaneAddOn = UpboundCrossplaneAddOn = UpboundCrossplaneAddOn_1 = __decorate([
    utils_1.supportsALL
], UpboundCrossplaneAddOn);
