// aws-cdk/lib/legacy-aws-auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SdkProvider = void 0;
exports.cached = cached;
exports.enableTracing = enableTracing;
const aws_auth_1 = require("./api/aws-auth");
const io_host_1 = require("./cli/io-host");
const singleton_plugin_host_1 = require("./cli/singleton-plugin-host");
/**
 * @deprecated
 */
function cached(obj, sym, fn) {
    if (!(sym in obj)) {
        obj[sym] = fn();
    }
    return obj[sym];
}
/**
 * Enable tracing in the CDK
 *
 * @deprecated cannot be enabled from outside the CDK
 */
function enableTracing(_enabled) {
    // noop
}
/**
 * @deprecated
 */
class SdkProvider {
    static async withAwsCliCompatibleDefaults(options = {}) {
        return aws_auth_1.SdkProvider.withAwsCliCompatibleDefaults({
            ...options,
            ioHelper: io_host_1.CliIoHost.instance().asIoHelper(),
            pluginHost: singleton_plugin_host_1.GLOBAL_PLUGIN_HOST,
        });
    }
    constructor(defaultCredentialProvider, defaultRegion, requestHandler = {}, logger) {
        return new aws_auth_1.SdkProvider(defaultCredentialProvider, defaultRegion, {
            pluginHost: singleton_plugin_host_1.GLOBAL_PLUGIN_HOST,
            ioHelper: io_host_1.CliIoHost.instance().asIoHelper(),
            requestHandler,
            logger,
        });
    }
}
exports.SdkProvider = SdkProvider;
