// @aws-quickstart/eks-blueprints/dist/addons/gateway-api-crds/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayApiCrdsAddOn = exports.GatewayApiCrdsBuilder = exports.GatewayApiCrdsStack = exports.GatewayCrdChannels = void 0;
// gateway-api-crds-stack.ts
const aws_cdk_lib_1 = require("aws-cdk-lib");
const utils_1 = require("../../utils");
const nested_stack_1 = require("../nested-stack");
exports.GatewayCrdChannels = {
    STANDARD: 'https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.2.0/standard-install.yaml',
    EXPERIMENTAL: 'https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.2.0/experimental-install.yaml',
};
const defaultProps = {
    gatewayApiCrdsUrl: exports.GatewayCrdChannels.STANDARD
};
class GatewayApiCrdsStack extends aws_cdk_lib_1.NestedStack {
    constructor(scope, id, props) {
        var _a;
        super(scope, id, props);
        this.manifestIds = [];
        if (!(props === null || props === void 0 ? void 0 : props.cluster)) {
            throw new Error("GatewayApiCrdsStack: 'cluster' must be provided in props.");
        }
        const cluster = props.cluster;
        const url = (_a = props === null || props === void 0 ? void 0 : props.gatewayApiCrdsUrl) !== null && _a !== void 0 ? _a : defaultProps.gatewayApiCrdsUrl;
        try {
            const yamlDocuments = (0, utils_1.loadExternalYaml)(url)
                .filter((manifest) => manifest && manifest.metadata && manifest.metadata.name);
            yamlDocuments.forEach((manifest) => {
                const manifestName = `sig-gateway-api-${manifest.metadata.name}`;
                return cluster.addManifest(manifestName, manifest);
            });
        }
        catch (error) {
            console.error('Error reading or parsing YAML file:', error);
            throw error;
        }
    }
}
exports.GatewayApiCrdsStack = GatewayApiCrdsStack;
class GatewayApiCrdsBuilder {
    constructor(props) {
        this.props = props;
        this.options = this.props;
    }
    build(scope, id, nestedStackProps) {
        return new GatewayApiCrdsStack(scope, id, {
            ...this.props,
            ...nestedStackProps
        });
    }
}
exports.GatewayApiCrdsBuilder = GatewayApiCrdsBuilder;
class GatewayApiCrdsAddOn extends nested_stack_1.NestedStackAddOn {
    constructor(props) {
        super({
            id: 'GatewayApiCrdsAddOn',
            builder: new GatewayApiCrdsBuilder(),
            nestedStackProps: props
        });
        this.options = props;
    }
    // Override the deploy method to inject cluster into props
    deploy(clusterInfo) {
        const cluster = clusterInfo.cluster; // Access cluster from clusterInfo
        if (!cluster) {
            throw new Error("Cluster is required for deployment.");
        }
        const updatedProps = {
            ...this.options,
            cluster,
        };
        // Get the builder for creating the stack
        const builder = new GatewayApiCrdsBuilder(updatedProps);
        const stack = cluster.stack;
        return Promise.resolve(builder.build(stack, 'gateway-api-crds-stack', updatedProps));
    }
}
exports.GatewayApiCrdsAddOn = GatewayApiCrdsAddOn;
