// @aws-quickstart/eks-blueprints/dist/addons/kube-proxy/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KubeProxyAddOn = void 0;
const aws_eks_1 = require("aws-cdk-lib/aws-eks");
const utils = require("../../utils");
const core_addon_1 = require("../core-addon");
const versionMap = new Map([
    [aws_eks_1.KubernetesVersion.V1_31, "v1.31.2-eksbuild.3"],
    [aws_eks_1.KubernetesVersion.V1_30, "v1.30.0-eksbuild.3"],
    [aws_eks_1.KubernetesVersion.V1_29, "v1.29.0-eksbuild.1"],
    [aws_eks_1.KubernetesVersion.V1_28, "v1.28.2-eksbuild.2"],
    [aws_eks_1.KubernetesVersion.V1_27, "v1.27.6-eksbuild.2"],
    [aws_eks_1.KubernetesVersion.V1_26, "v1.26.9-eksbuild.2"],
]);
const defaultProps = {
    addOnName: "kube-proxy",
    saName: "kube-proxy",
    versionMap: versionMap,
};
/**
 * Implementation of KubeProxy EKS add-on.
 */
let KubeProxyAddOn = class KubeProxyAddOn extends core_addon_1.CoreAddOn {
    deploy(clusterInfo) {
        return super.deploy(clusterInfo);
    }
    constructor(version, props) {
        super({
            version: version !== null && version !== void 0 ? version : "auto",
            ...defaultProps,
            ...props
        });
    }
};
exports.KubeProxyAddOn = KubeProxyAddOn;
__decorate([
    utils.conflictsWithAutoMode(null)
], KubeProxyAddOn.prototype, "deploy", null);
exports.KubeProxyAddOn = KubeProxyAddOn = __decorate([
    utils.supportsALL
], KubeProxyAddOn);
